### PR TITLE
Create semantic segmentation guide for DeepLab V3

### PR DESCRIPTION
**Pull request by sachinprasadhs**

### CODE REVIEW

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 1]
```diff
@@ -0,0 +1,874 @@
...
+Semantic segmentation is a type of computer vision task that involves assigning a
+class label such as person, bike, or background to each individual pixel of an
+image, effectively dividing the image into regions that correspond to different
+fobject classes or categories.
+
+![](https://miro.medium.com/v2/resize:fit:4800/format:webp/1*z6ch-2BliDGLIHpOPFY_Sw.png)
+
+
+
+KerasHub offers the DeepLabv3, DeepLabv3+, SegFormer etc models for semantic
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:23]: ", etc."

(I recommend running the text copy through Gemini and ask for typo fixes)

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 2]
```diff
@@ -0,0 +1,874 @@
...
+    if not override_extract and os.path.exists(data_directory):
+        logging.info("data directory %s already exist", data_directory)
+        return data_directory
+    data_file_path = keras.utils.get_file(
+        fname=fname, origin=data_url, cache_dir=cache_dir
+    )
+    # Extra the data into the same directory as the tar file.
+    data_directory = os.path.dirname(data_file_path)
+    logging.info("Extract data into %s", data_directory)
+    with tarfile.open(data_file_path) as f:
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:276]: Can you just use `extract = True` in `get_file`? I think everything in this function is covered by `get_file`.

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 3]
```diff
@@ -0,0 +1,874 @@
...
+        f.extractall(data_directory)
+    return os.path.join(data_directory, extracted_dir)
+
+
+def _parse_annotation_data(annotation_file_path):
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:281]: Not need for underscores -- this isn't a codebase, nothing is private

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 4]
```diff
@@ -0,0 +1,874 @@
...
+
+def _get_image_ids(data_dir, split):
+    data_file_mapping = {
+        "train": "train.txt",
+        "eval": "val.txt",
+        "trainval": "trainval.txt",
+        # TODO(tanzhenyu): add diff dataset
+        # "diff": "diff.txt",
+    }
+    with tf.io.gfile.GFile(
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:329]: Just use `open`

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 5]
```diff
@@ -0,0 +1,874 @@
...
+        "r",
+    ) as f:
+        image_ids = f.read().splitlines()
+        logging.info(f"Received {len(image_ids)} images for {split} dataset.")
+        return image_ids
+
+
+def _get_sbd_image_ids(data_dir, split):
+    data_file_mapping = {"sbd_train": "train.txt", "sbd_eval": "val.txt"}
+    with tf.io.gfile.GFile(
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:340]: Same here

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 6]
```diff
@@ -0,0 +1,874 @@
...
+        "segmentation/object/file_path",
+    ]
+    result = {}
+    for key in keys:
+        values = [value[key] for value in metadata]
+        result[key] = values
+    return result
+
+
+@tf.function(jit_compile=True)
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:444]: No need

**Review by fchollet — COMMENTED** [review]
Thanks for the PR!

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py (hunk 2 above)
Comment (sachinprasadhs) [reply guides/keras_hub/semantic_segmentation_deeplab_v3.py]: Done, Thanks for the suggestion.

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 7]
```diff
@@ -0,0 +1,845 @@
...
+    outputs = outputs.batch(4, drop_remainder=True)
+    return outputs
+
+
+train_ds = preprocess_inputs(train_ds)
+batch = train_ds.take(1).get_single_element()
+
+"""
+A batch of this preprocessed input training data can be visualized using the
+`plot_images_masks` function. This function takes a
```
Comment (divyashreepathihalli) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:637]: looks like this line length is much shorter than the line limit before going to next line. Update the text here to be just under the line limit before going to next line

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 8]
```diff
@@ -0,0 +1,845 @@
...
+                    "bbox": [ymin, xmin, ymax, xmax],
+                    "is_truncated": is_truncated,
+                    "is_difficult": is_difficult,
+                }
+            )
+
+        return {"width": width, "height": height, "objects": objects}
+
+
+def get_image_ids(data_dir, split):
```
Comment (divyashreepathihalli) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:281]: add a short comment to describe what each helper function is doing.

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 9]
```diff
@@ -0,0 +1,845 @@
...
+"""
+The preprocessing is applied to the evaluation dataset `eval_ds`.
+"""
+eval_ds = preprocess_inputs(eval_ds)
+
+"""
+## Data Augmentation
+
+Keras provides a variety of image augmentation options. In this example, we will
+use the `RandomFlip` augmentation to augment the training dataset. The
```
Comment (divyashreepathihalli) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:682]: I want to make sure that the segmentation masks are also flipped in the training dataset correct?

Comment (sachinprasadhs) [reply guides/keras_hub/semantic_segmentation_deeplab_v3.py:682]: Yes, `train_ds` is having both images and batch, in the map function it flips both images and masks.
flipped images are visualized in the function here plot_images_masks(batch["images"], batch["segmentation_masks"])

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 10]
```diff
@@ -0,0 +1,845 @@
+"""
+Title: Semantic Segmentation with KerasHub
+Author: [Sachin Prasad](https://github.com/sachinprasad)
```
Comment (divyashreepathihalli) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:3]: If this guide was adapted from KerasCV guide please also mention the original authors here.

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py (hunk 8 above)
Comment (sachinprasadhs) [reply guides/keras_hub/semantic_segmentation_deeplab_v3.py:281]: Done

**Comment by divyashreepathihalli** [comment]
Thanks @sachinprasadhs can you please generate the rest of the files?

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 11]
```diff
@@ -0,0 +1,861 @@
...
+backbones to achieve accurate and detailed semantic segmentation.
+
+DeepLabv3+ extends DeepLabv3 by adding a simple yet effective decoder module to
+refine the segmentation results, especially along object boundaries. Both models
+have achieved state-of-the-art results on a variety of image segmentation
+benchmarks.
+
+### References
+[Encoder-Decoder with Atrous Separable Convolution for Semantic Image
+Segmentation](https://arxiv.org/abs/1802.02611)<br>
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:38]: Links should be on a single line. Also, remove br tags

File: guides/keras_hub/semantic_segmentation_deeplab_v3.py [hunk 12]
```diff
@@ -0,0 +1,861 @@
...
+"""shell
+pip install -q --upgrade keras-hub
+pip install -q --upgrade keras
+"""
+
+"""
+After installing `keras` and `keras-hub`, set the backend for `keras`.
+This guide can be run with any backend (Tensorflow, JAX, PyTorch).
+
+```
```
Comment (fchollet) [inline guides/keras_hub/semantic_segmentation_deeplab_v3.py:63]: You can remove this block -- it is already in the code.

**Comment by divyashreepathihalli** [comment]
@sachinprasadhs Any updated on this?

File: guides/md/keras_hub/semantic_segmentation_deeplab_v3.md [hunk 13]
```diff
@@ -0,0 +1,11901 @@
...
+
+<div class="k-default-codeblock">
+```
+/home/sachinprasad/projects/env/lib/python3.11/site-packages/keras/src/models/functional.py:225: UserWarning: The structure of `inputs` doesn't match the expected structure: ['keras_tensor']. Received: the structure of inputs=*
+  warnings.warn(
+
+```
+</div>
+    
+![png](/home/sachinprasad/projects/keras-io/guides/img/semantic_segmentation_deeplab_v3/semantic_segmentation_deeplab_v3_9_1.png)
```
Comment (divyashreepathihalli) [inline guides/md/keras_hub/semantic_segmentation_deeplab_v3.md:153]: there is some issue with autogen. The file paths need to updated to be relative here

Comment (sachinprasadhs) [reply guides/md/keras_hub/semantic_segmentation_deeplab_v3.md]: Made changes manually, could you please check.

**Review by fchollet — APPROVED** [review]
